pkg/internal/itest/client_retrieval_test.go

```diff
@@ -0,0 +1,281 @@
+}
+
+func eventSliceFilter(events []datatransfer.Event, code datatransfer.EventCode) []datatransfer.Event {
+	filtered := make([]datatransfer.Event, 0)
+	for _, event := range events {
+		if event.Code == code {
+			filtered = append(filtered, event)
+		}
+	}
+	return filtered
+}
+
+type mockRetrievalNet struct {
+	remoteEvents []datatransfer.Event
+
+	mn               mocknet.Mocknet
+	hostLocal        host.Host
+	hostRemote       host.Host
+	linkSystemRemote linking.LinkSystem
+}
+
+func newMockRetrievalNet() *mockRetrievalNet {
+	mqn := &mockRetrievalNet{
+		remoteEvents: make([]datatransfer.Event, 0),
+	}
+	return mqn
+}
+
+func (mrn *mockRetrievalNet) setup(ctx context.Context) error {
+	// Setup remote datastore and blockstore
+	dsRemote := dss.MutexWrap(datastore.NewMapDatastore())
+	dtDsRemote := namespace.Wrap(dsRemote, datastore.NewKey("datatransfer"))
+	bsRemote := bstore.NewBlockstore(namespace.Wrap(dsRemote, datastore.NewKey("blockstore")))
+	mrn.linkSystemRemote = storeutil.LinkSystemForBlockstore(bsRemote)
+
+	// Setup network
+	mrn.mn = mocknet.New()
+	var err error
+	mrn.hostRemote, err = mrn.mn.GenPeer()
+	if err != nil {
+		return err
+	}
+	mrn.hostLocal, err = mrn.mn.GenPeer()
+	if err != nil {
+		return err
+	}
+	if err := mrn.mn.LinkAll(); err != nil {
+		return err
+	}
+
+	// Setup remote data transfer
+	gsNetRemote := gsnet.NewFromLibp2pHost(mrn.hostRemote)
+	dtNetRemote := dtnet.NewFromLibp2pHost(mrn.hostRemote, dtnet.RetryParameters(0, 0, 0, 0))
+	gsRemote := gsimpl.New(ctx, gsNetRemote, mrn.linkSystemRemote)
+	gstpRemote := gstransport.NewTransport(mrn.hostRemote.ID(), gsRemote)
+	dtRemote, err := dtimpl.NewDataTransfer(dtDsRemote, dtNetRemote, gstpRemote)
+	if err != nil {
+		return err
+	}
+
+	// Wait for remote data transfer to be ready
+	ready := make(chan error, 1)
+	dtRemote.OnReady(func(err error) {
+		ready <- err
+	})
+	if err := dtRemote.Start(ctx); err != nil {
+		return err
+	}
+	select {
+	case <-ctx.Done():
+	case err := <-ready:
+		if err != nil {
+			return err
+		}
+	}
+
+	// Register DealProposal voucher type with automatic Pull acceptance
+	remoteDealValidator := &mockDealValidator{acceptPull: true}
+	if err := dtRemote.RegisterVoucherType(retrievalmarket.DealProposalType, remoteDealValidator); err != nil {
+		return err
+	}
+
+	// Record remote events
+	subscriberRemote := func(event datatransfer.Event, channelState datatransfer.ChannelState) {
+		mrn.remoteEvents = append(mrn.remoteEvents, event)
+	}
+	dtRemote.SubscribeToEvents(subscriberRemote)
+
+	return nil
+}
+
+func (mrn *mockRetrievalNet) teardown() error {
+	return mrn.mn.Close()
+}
+
+func (mrn *mockRetrievalNet) generateRemoteUnixFSFile() (cid.Cid, []byte, error) {
+	// a file of 4MiB random bytes, packaged into unixfs DAGs, stored in the remote blockstore
+	delimited := io.LimitReader(rand.Reader, 1<<22)
+	buf := new(bytes.Buffer)
+	delimited = io.TeeReader(delimited, buf)
```

Comment:
minor: Growing the buffer would reduce allocs when size is know.
Also you might want to define `const ` for `4 << 20` if you decide to reflect on this suggestion.

```suggestion
	buf.Grow(4 << 20)
	delimited = io.TeeReader(delimited, buf)
```